### PR TITLE
Source Oracle DB: Create comprehensive data type tests

### LINKED ISSUE #3796: source Oracle DB: Create comprehensive data type tests
## Tell us about the problem you're trying to solve
Same as https://github.com/airbytehq/airbyte/issues/3562 but for Oracle

Version for datatypes 11g of the dialect.
https://docs.oracle.com/cd/E11882_01/server.112/e41085/sqlqr06002.htm#SQLQR959

TODO:

1. Read documentation and prepare tricky use cases for each type of data. 
2. Implement comprehensive test. 
3. Prepare pull request, pass all checks. 

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200368060087111/1200413427486564) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by heade**
## What
Ticket: https://github.com/airbytehq/airbyte/issues/3796 : source Oracle DB: Create comprehensive data type tests
We are creating comprehensive integration tests for the Oracle source connector. Such a test guarantees that connector implementation is infallible even in cases when the source contains unexpected/extreme values.

## How
Prepare a list of cases for testing in issue https://github.com/airbytehq/airbyte/issues/3796
Implement the abstract `SourceAbstractTest` test with all cases from the issue.

## Recommended reading order
1. `x.java`

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [x] Issue acceptance criteria met
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in output spec
- [x] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Documentation updated 
    - [ ] README
    - [ ] CHANGELOG.md
    - [ ] Reference docs in the `docs/integrations/` directory.
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

### CODE REVIEW

**Review by DoNotPanicUA — APPROVED** [review]

File: airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    addDataTypeTestData(
+        TestDataHolder.builder()
+            .sourceType("VARCHAR2")
+            .airbyteType(JsonSchemaPrimitive.STRING)
+            .fullSourceDataType("VARCHAR2(256)")
+            .addInsertValues("chr(33) || chr(34) || chr(35) || chr(36) || chr(37) || chr(38) || chr(39) || chr(40) || chr(41)")
+            .addExpectedValues("!\"#$%&'()")
+            .build());
+
+    addDataTypeTestData(
```
Comment (yaroslav-hrytsaienko) [inline airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java:133]: NCHAR  type can be added to verification within max lengs and zero length value. The same for NVARCHAR2

File: airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        TestDataHolder.builder()
+            .sourceType("BLOB")
+            .airbyteType(JsonSchemaPrimitive.STRING)
+            .addInsertValues("utl_raw.cast_to_raw('some content here')", "null")
+            .addExpectedValues("c29tZSBjb250ZW50IGhlcmU=", null)
+            .build());
+
+    addDataTypeTestData(
+        TestDataHolder.builder()
+            .sourceType("CLOB")
```
Comment (yaroslav-hrytsaienko) [inline airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java:237]: what about NCLOB type?

File: airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        TestDataHolder.builder()
+            .sourceType("DATE")
+            .airbyteType(JsonSchemaPrimitive.STRING)
+            .addInsertValues("to_date('-4700/01/01','syyyy/mm/dd')", "to_date('9999/12/31 23:59:59','yyyy/mm/dd hh24:mi:ss')", "null")
+            .addExpectedValues("4700-01-01T00:00:00Z", "9999-12-31T23:59:59Z", null)
+            .build());
+
+    addDataTypeTestData(
+        TestDataHolder.builder()
+            .sourceType("TIMESTAMP")
```
Comment (yaroslav-hrytsaienko) [inline airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java:202]: DATETIME, TIMESTAMP WITH LOCAL TIME ZONE, TIMESTAMP WITH TIME ZONE datatype?

File: airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+    addDataTypeTestData(
+        TestDataHolder.builder()
+            .sourceType("CLOB")
+            .airbyteType(JsonSchemaPrimitive.STRING)
+            .addInsertValues("utl_raw.cast_to_raw('some content here')", "null")
+            .addExpectedValues("736F6D6520636F6E74656E742068657265", null)
+            .build());
+
+    addDataTypeTestData(
```
Comment (yaroslav-hrytsaienko) [inline airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java:243]: BFILE, LONG RAW, XMLType?

**Review by yaroslav-hrytsaienko — COMMENTED** [review]
@heade please look at missing datatypes, and create tests for them also. As well corner cases with values can be extended, like maximum length and empty values. 
Another aspecs is that i didn't find the ticket description as a table with all use cases for the test. Can you please create such ? Please look like it's done for https://github.com/airbytehq/airbyte/issues/3562#issuecomment-850838160

File: airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java (hunk 3 above)
Comment (heade) [reply airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java:202]: @yaroslav-hrytsaienko there is no DATETIME datatype. Only DATE, TIMESTAMP WITH LOCAL TIME ZONE, TIMESTAMP WITH TIME ZONE. The last one you can see below. Thanks

File: airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.io.airbyte.integration_tests.sources;
```
Comment (subodh1810) [inline airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java:25]: The package doesnt seem right. If you take a look at `MySqlSourceComprehensiveTest`, the package is ` io.airbyte.integrations.source.mysql;` so I would expect similar pattern here as well

File: airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+            .addInsertValues("utl_raw.cast_to_raw('some content here')", "null")
+            .addExpectedValues("c29tZSBjb250ZW50IGhlcmU=", null)
+            .build());
+
+    addDataTypeTestData(
+        TestDataHolder.builder()
+            .sourceType("LONG")
+            .airbyteType(JsonSchemaPrimitive.STRING)
+            .fullSourceDataType("LONG RAW")
+            // @TODO stream fails when reading data back
```
Comment (subodh1810) [inline airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java:285]: Can you create a new issue to fix the implementation to handle these tests

File: airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+            .addExpectedValues("2009-02-21 18:00:00.0 EST", "2009-02-21 18:00:00.0 -5:00")
+            .build());
+
+    DateFormat utcFormat = new SimpleDateFormat("dd-MMM-yyyy HH:mm:ss");
+    utcFormat.setTimeZone(TimeZone.getTimeZone(Calendar.getInstance().getTimeZone().getID()));
+    Date date = null;
+    try {
+      date = utcFormat.parse("21-Feb-2009 18:00:00");
+    } catch (ParseException e) {
+      e.printStackTrace();
```
Comment (subodh1810) [inline airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java:232]: Why just print stacktrace?

**Review by subodh1810 — COMMENTED** [review]
Can you trigger the test in the CI to make sure they pass here as well.

File: airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java (hunk 6 above)
Comment (heade) [reply airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/OracleSourceComprehensiveTest.java]: https://github.com/airbytehq/airbyte/issues/4307

**Review by subodh1810 — APPROVED** [review]

**Comment by sherifnada** [comment]
@heade anything blocking us from merging this?

**Comment by DoNotPanicUA** [comment]
> @heade anything blocking us from merging this?

yeah, he is on sick leave. :sneezing_face:
so, I've merged the PR :blush:

**Pull request merged**